sunpy/map/mapbase.py

```diff
@@ -1963,50 +1963,122 @@ def draw_grid(self, axes=None, grid_spacing: u.deg = 15*u.deg, annotate=True, **
         -----
-        Keyword arguments are passed onto `matplotlib.patches.Circle`.
+        Keyword arguments are passed onto the patches.
+
+        If the limb is a true circle, ``visible`` will instead be
+        `~matplotlib.patches.Circle` and ``hidden`` will be ``None``.
+
+        To avoid trigger Matplotlib auto-scaling, these patches are added as generic
+        artists instead of patches.  One consequence is that the plot legend is not
+        populated automatically when the limb is specified with a text label.
         """
         # Put import here to reduce sunpy.map import time
         from matplotlib import patches
+        from matplotlib.path import Path
 
         # Don't use _check_axes() here, as drawing the limb works fine on none-WCSAxes,
         # even if the image is rotated relative to the axes
         if not axes:
             axes = wcsaxes_compat.gca_wcs(self.wcs)
+        is_wcsaxes = wcsaxes_compat.is_wcsaxes(axes)
 
-        transform = wcsaxes_compat.get_world_transform(axes)
-        if wcsaxes_compat.is_wcsaxes(axes):
-            radius = self.rsun_obs.to(u.deg).value
-        else:
-            radius = self.rsun_obs.value
-        c_kw = {'radius': radius,
-                'fill': False,
+        c_kw = {'fill': False,
                 'color': 'white',
-                'zorder': 100,
-                'transform': transform
-                }
+                'zorder': 100}
         c_kw.update(kwargs)
 
-        circ = patches.Circle([0, 0], **c_kw)
-        axes.add_artist(circ)
+        # Obtain the solar radius and the world->pixel transform
+        if not is_wcsaxes:
+            radius = self.rsun_obs.value
+            transform = axes.transData
+        else:
+            radius = self.rsun_obs.to_value(u.deg)
 
-        return [circ]
+            if isinstance(self.coordinate_frame, Helioprojective):
+                frame = self.coordinate_frame
+            else:
+                frame = Helioprojective(observer=self.observer_coordinate,
+                                        obstime=self.date, rsun=self.rsun_meters)
+            transform = axes.get_transform(frame)
+
+        # transform is always passed on as a keyword argument
+        c_kw.setdefault('transform', transform)
+
+        # If not WCSAxes or if the map's frame matches the axes's frame, we can use Circle
+        if not is_wcsaxes or frame == axes._transform_pixel2world.frame_out:
+            c_kw.setdefault('radius', radius)
+
+            circ = patches.Circle([0, 0], **c_kw)
+            axes.add_artist(circ)
+
+            return circ, None
+
+        # Otherwise, we use Polygon to be able to distort the limb
+
+        theta = np.linspace(0, 2*np.pi, resolution+1)[:-1] << u.rad
+        # Use the Cartesian approximation of helioprojective coordinates because the Sun is small
+        Tx, Ty = radius * np.cos(theta), radius * np.sin(theta)
+
+        # Perform the round-trip transformation world->pixel->world
```

Comment:
```suggestion
        # Perform the round-trip transformation map_world->map_pixel->axes_world
```
Or similar, to be clear that this is to transform from the map coordinates to the coordinates used for the axes.

Reply:
It isn't.  In full detail, the round-trip transformation is `map_world(2D)`->`map_world(3D)`->`axes_world(3D)`->`axes_world(2D)`->`axes_pixel`->`axes_world(2D)`->`axes_world(3D)`->`map_world(3D)`->`map_world(2D)`.

Reply:
Ah, so to make sure I understand this essentially:
- Projects all the 3D limb coordinates on to the axes-observer-facing side of the Sun
- Checks if any of the projected coordinates have moved, and if so labels them as far-side

I presume there's good reasons for doing this instead of transforming to axes-HPC, and testing `transformed_coord.distance >= coord.rsun`? I think this way would be easier to follow in the code, and possibly involve less transforms so be better performance, but may have other issues I haven't thought of.

Reply:
Moot